### PR TITLE
Bayes nn

### PR DESCRIPTION
**Pull request by bdtmnk**

### CODE REVIEW

**Comment by aperloff** [comment]
How far along are you with this? Should I review it or is it more of a work in progress? Also, please add a description to your PR.

**Comment by bdtmnk** [comment]
Dear Alexx, 

There is more work will be added. 
I will inform you at which point to review. 

Best regards, 
Leonid Didukh

File: content/bayes_nn/bayes.md [hunk 1]
```diff
@@ -0,0 +1,318 @@
+# Bayesian Neural Network
+
+The usual Neural Network are optimized in way to get fixed value of weights and biases that allows the model perform specific task successfully. Instead in
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:3]: "Usually, neural networks are optimized in order to get a fixed value for the weights and biases which allow the model perform a specific task successfully."

File: content/bayes_nn/bayes.md [hunk 2]
```diff
@@ -0,0 +1,318 @@
+# Bayesian Neural Network
+
+The usual Neural Network are optimized in way to get fixed value of weights and biases that allows the model perform specific task successfully. Instead in 
+Bayesian Neural Network the weights and biases are the distribution, this type of model could be treated as a ensemble of many neural networks trained by the Bayesian inference.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:4]: "In a Bayesian neural network the weights and biases are distributed rather than fixed. This type of model could be treated as an ensemble of many neural networks, train using a Bayesian inference."

File: content/bayes_nn/bayes.md [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+Bayesian approach for the neural networks allows to estimate the uncertainty and make the decision of the model more robust according to the input data.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:6]: "Using a Bayesian approach for the neural network training allows the analyzer to estimate the uncertainty and to make the decision of the model more robust against the input data."

File: content/bayes_nn/bayes.md [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+
+### Difference between usual NN and BNN
+
+
+![Placeholder](../images/bayes_nn/diff.png)
+
+
+### Training of NN and BNN
+=== "NN"
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:17]: Since the title is in the images, is this notation really needed?

File: content/bayes_nn/bayes.md [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    ![Placeholder](../images/bayes_nn/trainingNN.png)
+    The parameters  ![formula](https://render.githubusercontent.com/render/math?math=\theta ) are optimized in order to minimaze the loss function
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:19]: The formula after "parameters" isn't being rendered.

"minimaze" -> "minimize"

Put a "." at the end of the sentence.

File: content/bayes_nn/bayes.md [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+=== "BNN"
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:21]: Again, is this notation needed?

File: content/bayes_nn/bayes.md [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    ![Placeholder](../images/bayes_nn/bayesNN.png)
+     The process is to learn the probability distributions for weights and biases that maximize the likelihood of getting a high probability for the correct data/label ![formula](https://render.githubusercontent.com/render/math?math=D(x,y) ) pairs.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:23]: The formula in the middle of the sentence isn't being rendered.

File: content/bayes_nn/bayes.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    The parameters of the weights distributions - mean and standart deviation are the product of the loss function optimization
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:24]: "The parameters of the weight distributions, mean and standard deviation, are the results of the loss function optimization."

File: content/bayes_nn/bayes.md [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+#### Training Procedure
+    1. Introduce the prior distribution over model parameter w
+    2. Compute posterio p(w|D) using Bayesian rule
+    3. Make the average over the posterior distribution
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:29]: "Make" -> "Take"

File: content/bayes_nn/bayes.md [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+
+
+### Prediction of NN and BNN
+=== "NN"
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:34]: Maybe '#### NN' rather than what is there?

File: content/bayes_nn/bayes.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+    ![Placeholder](../images/bayes_nn/PredictionNN.png)
+
+=== "BNN"
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:37]: See previous comment.

File: content/bayes_nn/bayes.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    ![Placeholder](../images/bayes_nn/PredictionBNN.png)
+
+
+### Uncertainty
+The uncertainty that are quatified by the BNN are categorized in the next way:
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:42]: "There are two types of BNN uncertainties:
1. Alletonic - uncertainties due to the lack of knowledge, comes from data or environment formula )
2. Epistemic - uncertainties of the model parameter formula)
"

Also, the formulas are not being rendered.

File: content/bayes_nn/bayes.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+
+
+
+
+
+## Packages
+The are several packages for the probabilistic neural network, the tensorflow probability and pyro are the most consistent
+
+=== "Tensorflow"
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:59]: I don't think the `===` are doing what you think they are.

File: content/bayes_nn/bayes.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+    ![formula](https://render.githubusercontent.com/render/math?math=p(y|x,\theta))
+
+
+
+
+
+
+
+## Packages
+The are several packages for the probabilistic neural network, the tensorflow probability and pyro are the most consistent
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:57]: Here we will list a few of the machine learning packages which can be used to develop a probabilistic neural network.

File: content/bayes_nn/bayes.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+## Modules Description:
+
+### Distribution and sampling
+
+=== "Tensorflow"
+
+=== "Pyro"
+
+### Distribution and sampling
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:77]: Duplicate section?

File: content/bayes_nn/bayes.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+
+=== "Tensorflow"
+
+=== "Pyro"
+
+
+
+
+Let's consider simple linear regression as an example and compare it to the bayesian analog.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:86]: Capitalize the word Bayesian.

File: content/bayes_nn/bayes.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+The output of the model:
+
+![Placeholder](../images/bayes_nn/lr.png)
+
+
+
+
+## Variational Autoencoder
+
+The generative models could be build using the bayesian neural network.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:201]: "Generative models can be built using a Bayesian neural network. The variational autoencoder is one popular way to forma generative model."

File: content/bayes_nn/bayes.md [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+The Variantional Autoencoder is the popular way for data synthethis.
+
+Let's consider the example of generating the images:
+
+The generating process consist of two steps:
+
+1. Samling the latent variable from prior distribution
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:208]: "Samling" -> "Sampling"

File: content/bayes_nn/bayes.md [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+Objective:
+
+![formula](https://render.githubusercontent.com/render/math?math=p(z)) the prior on the latent representation ![formula](https://render.githubusercontent.com/render/math?math=z) ,
+![formula](https://render.githubusercontent.com/render/math?math=q(z|x)), the variational encoder, and
+![formula](https://render.githubusercontent.com/render/math?math=p(x|z)), the decoder — how likely is the image x given the latent representation z.
+
+### Loss
+
+Once we define the procedure for the generation process the Objective function should be chosen for the optimization process. In order to train the network, we maximize the ELBO (Evidence Lower Bound) objective.
```
Comment (aperloff) [inline content/bayes_nn/bayes.md:220]: "Objective" -> "objective"

**Review by aperloff — COMMENTED** [review]
This is a preliminary set of review comments. It seems that this document still needs some work.

**Comment by aperloff** [comment]
@bdtmnk Can you please update this PR to fix the conflict and apply the requested changes?

**Comment by aperloff** [comment]
Closing this PR for lack of progress. While this PR is being closed, we feel that documenting Bayesian Neural Networks is important. Therefore the work has been rebased in PR #30 and work will continue there.
